src/wasm/kotlin/wasm/arcs/Particle.kt

```diff
@@ -48,22 +49,37 @@ abstract class Particle : WasmObject() {
         onHandleSync(handle, toSync.isEmpty())
     }
 
-    fun renderSlot(slotName: String, content: String) {
-        render(slotName.toWasmString(), content.toWasmString())
-    }
-
     abstract fun onHandleUpdate(handle: Handle)
     abstract fun onHandleSync(handle: Handle, willSync: Boolean)
-    abstract fun requestRender(slotName: String)
+    fun renderSlot(slotName: String, sendTemplate: Boolean = true, sendModel: Boolean = true) {
+      val template = if (sendTemplate) getTemplate(slotName) else ""
+      var model = ""
+      if (sendModel) {
+        val sb = StringBuilder()
+        var i = 0
+        populateModel(slotName) { key: String, value: String ->
+          sb.append(key.length).append(":").append(key)
+          sb.append(value.length).append(":").append(value)
+          i++
+        }
+        model = "$i:$sb"
+      }
+
+      render(this.toWasmAddress(), slotName.toWasmString(), template.toWasmString(), model.toWasmString())
+    }
 
     open fun fireEvent(slotName: String, eventName: String) {
-        eventHandlers[eventName]?.invoke()
-        requestRender(slotName)
+      eventHandlers[eventName]?.invoke()
+      renderSlot(slotName)
     }
+
+    open fun getTemplate(slotName: String): String = ""
+    open fun populateModel(slotName: String, accumulateModel: (String, String) -> Unit): String = ""
 }
 
 abstract class Handle : WasmObject() {
     var name: String? = null
+    var particle_: Particle? = null
```

Comment:
Is the '_' suffix a standard (/ our standard)? for internal data?

Reply:
I was following a convention from cpp. There's no precedent in the file for that, so I'll drop the `_`.